src/test/beatstest.cpp

```diff
@@ -0,0 +1,458 @@
+    }
+    EXPECT_DOUBLE_EQ(
+            pos.getValue(), m_pBeats1->getLastBeatPosition().getValue());
+
+    // Past end
+    auto iter2 = m_pBeats1->findBeats(m_pBeats1->getFirstBeatPosition(),
+            FramePos(
+                    m_pBeats1->getLastBeatPosition().getValue() + 10000000000));
+    while (iter2->hasNext()) {
+        auto beat = iter2->next();
+        pos = FramePos(beat.getFramePosition().getValue());
+        EXPECT_TRUE(pos.getValue());
+    }
+    EXPECT_DOUBLE_EQ(
+            pos.getValue(), m_pBeats1->getLastBeatPosition().getValue());
+
+    // Before begining
+    auto iter3 = m_pBeats1->findBeats(
+            FramePos(m_pBeats1->getFirstBeatPosition().getValue() - 1000000),
+            m_pBeats1->getLastBeatPosition());
+    EXPECT_DOUBLE_EQ(iter3->next().getFramePosition().getValue(),
+            m_pBeats1->getFirstBeatPosition().getValue());
+    while (iter3->hasNext()) {
+        auto beat = iter3->next();
+        pos = FramePos(beat.getFramePosition().getValue());
+        EXPECT_TRUE(pos.getValue());
+    }
+    EXPECT_DOUBLE_EQ(pos.getValue(), m_pBeats1->getLastBeatPosition().getValue());
+}
+
+TEST_F(BeatsTest, Translate) {
+    FrameDiff_t delta = 500;
+
+    // Move the grid delta frames
+    m_pBeats1->translate(delta);
+
+    // All beats must have been displaced by delta frames
+    auto iter1 = m_pBeats1->findBeats(m_pBeats1->getFirstBeatPosition(),
+            m_pBeats1->getLastBeatPosition());
+    auto iter2 = m_pBeats2->findBeats(m_pBeats2->getFirstBeatPosition(),
+            m_pBeats2->getLastBeatPosition());
+    while (iter1->hasNext()) {
+        double pos1 = iter1->next().getFramePosition().getValue();
+        double pos2 = iter2->next().getFramePosition().getValue();
+        EXPECT_DOUBLE_EQ(pos1, pos2 + delta);
+    }
+    //    EXPECT_EQ(iter1->hasNext(), iter2->hasNext());
+}
+
+TEST_F(BeatsTest, FindClosest) {
+    // Test deltas ranging from previous beat to next beat
+    for (FrameDiff_t delta = -m_iSampleRate; delta <= m_iSampleRate; delta++) {
+        auto iter1 = m_pBeats1->findBeats(m_pBeats1->getFirstBeatPosition(),
+                m_pBeats1->getLastBeatPosition());
+        while (iter1->hasNext()) {
+            FramePos pos = FramePos(iter1->next().getFramePosition().getValue());
+            FramePos foundPos = m_pBeats1->findClosestBeat(pos + delta);
+            // Correct change of beat
+            FramePos expectedPos = pos +
+                    (delta > (m_iSampleRate / 2.0) ? m_iSampleRate : 0) +
+                    (delta < (-m_iSampleRate / 2.0) ? -m_iSampleRate : 0);
+            // Enforce boundaries
+            expectedPos = std::min(expectedPos, m_pBeats1->getLastBeatPosition());
+            expectedPos = std::max(expectedPos, m_pBeats1->getFirstBeatPosition());
+            EXPECT_DOUBLE_EQ(foundPos.getValue(), expectedPos.getValue());
+        }
+        break;
+    }
+}
+
+TEST_F(BeatsTest, ChangingTimeSignatureShouldNotChangeBpm) {
```

Comment:
```
 7/13 Test #45: BeatsTest.ChangingTimeSignatureShouldNotChangeBpm ...***Failed    0.21 sec
faad2::LibLoader - Successfully loaded FAAD2 library "libfaad.so" version 2.9.2
Note: Google Test filter = BeatsTest.ChangingTimeSignatureShouldNotChangeBpm
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from BeatsTest
[ RUN      ] BeatsTest.ChangingTimeSignatureShouldNotChangeBpm
DEBUG ASSERT: "m_beatsProto.time_signature_markers_size() > 0" in function void mixxx::BeatsInternal::generateBeatsFromMarkers() at /home/jan/Projects/mixxx/src/track/beats.cpp:829
DEBUG ASSERT: "m_beatsProto.time_signature_markers_size() > 0" in function void mixxx::BeatsInternal::generateBeatsFromMarkers() at /home/jan/Projects/mixxx/src/track/beats.cpp:829
/home/jan/Projects/mixxx/src/test/beatstest.cpp:456: Failure
Expected equality of these values:
  oldBpm
    Which is: 8-byte object <00-00 00-00 00-00 4E-40>
  newBpm
    Which is: 8-byte object <00-00 00-00 00-00 5E-40>
[  FAILED  ] BeatsTest.ChangingTimeSignatureShouldNotChangeBpm (1 ms)
[----------] 1 test from BeatsTest (1 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (1 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] BeatsTest.ChangingTimeSignatureShouldNotChangeBpm

 1 FAILED TEST
```

Reply:
I am working on this. This is a new test which is needed since we can change beat length now by changing the time signature. So the BPM calculation algorithm will be changed accordingly.